plugins/web-worker-offloading/hooks.php

```diff
@@ -0,0 +1,146 @@
+		'lib'     => substr( $plugin_dir, strpos( $plugin_dir, $content_dir_basename ) - 1 ) . 'build/',
+		'forward' => array(),
+	);
+
+	/**
+	 * Add configuration for PartyTown.
+	 *
+	 * @since n.e.x.t
+	 * @see <https://partytown.builder.io/configuration>.
+	 * @param array $config Configuration for PartyTown.
+	 * @return array
+	 */
+	return apply_filters( 'wwo_configuration', $config );
+}
+
+/**
+ * Initialize Web Worker Offloading.
+ *
+ * @since n.e.x.t
+ */
+function wwo_init(): void {
+	$partytown_js = file_get_contents( __DIR__ . '/build/partytown.js' ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents -- It's a local filesystem path not a remote request.
+
+	if ( ! $partytown_js ) {
+		return;
+	}
+
+	wp_register_script(
+		'partytown',
+		'',
+		array(),
+		WEB_WORKER_OFFLOADING_VERSION,
+		array( 'in_footer' => false )
+	);
+
+	wp_add_inline_script(
+		'partytown',
+		sprintf(
+			'window.partytown = %s;',
+			wp_json_encode( wwo_configuration() )
+		),
+		'before'
+	);
+
+	wp_add_inline_script(
+		'partytown',
+		$partytown_js,
+		'after'
+	);
+}
+add_action( 'wp_enqueue_scripts', 'wwo_init' );
+
+/**
+ * Helper function to get all scripts tags which has `partytown` dependency.
+ *
+ * @since n.e.x.t
+ *
+ * @return string[] Array of script handles.
+ */
+function wwo_get_partytown_handles(): array {
+	/**
+	 * Array of script handles which has `partytown` dependency.
+	 *
+	 * @var string[]
+	 */
+	static $partytown_handles = array();
+
+	if ( ! empty( $partytown_handles ) ) {
+		return $partytown_handles;
+	}
+
+	foreach ( wp_scripts()->registered as $handle => $script ) {
+		if ( ! empty( $script->deps ) && in_array( 'partytown', $script->deps, true ) ) {
+			$partytown_handles[] = $handle;
+		}
+	}
+
+	return $partytown_handles;
+}
+
+/**
+ * Update script type for handles having `partytown` as dependency.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $tag Script tag.
+ * @param string $handle Script handle.
+ * @param string $src Script source.
+ *
+ * @return string $tag Script tag with type="text/partytown".
+ */
+function wwo_update_script_type( string $tag, string $handle, string $src ): string {
+	$partytown_handles = wwo_get_partytown_handles();
+
+	if ( in_array( $handle, $partytown_handles, true ) ) {
+		$before_script = wp_scripts()->get_inline_script_data( $handle, 'before' );
+		$after_script  = wp_scripts()->get_inline_script_data( $handle, 'after' );
+
+		if ( ! empty( $before_script ) || ! empty( $after_script ) ) {
+			_doing_it_wrong(
+				'wp_add_inline_script',
+				sprintf(
+					/* translators: %s: script handle */
+					esc_html__( 'Cannot add inline script "%s" to scripts with a "partytown" dependency. Script will continue to load in the main thread.', 'web-worker-offloading' ),
+					'<a href="https://developer.wordpress.org/reference/functions/wp_add_inline_script/">wp_add_inline_script()</a>'
+				),
+				esc_html( WEB_WORKER_OFFLOADING_VERSION )
+			);
```

Comment:
Can I get some feedback on this part? It doesn't offload a script to a web worker if it has dependent scripts that load before or after it. My concern is that doing so might disrupt the loading order or break some dependencies.

Reply:
It should be fine to have a before script, but an after script would indeed be a problem.

Reply:
What happens if the `before` script is async? Partytown searches for scripts with `type="text/partytown"` once `DOMContentLoaded` is triggered. If the before script is `async` and still downloading, it won't block `DOMContentLoaded`, allowing Partytown to start executing the scripts in the worker thread.

Reply:
Inline scripts cannot be `async` in general. The only way to _defer_ an inline script would be to make it a module, but this isn't supported by the `wp_add_inline_script()` API.

With the introduction of the script loading strategies, there is this [`filter_eligible_strategies()`](https://github.com/WordPress/wordpress-develop/blob/2c818502ea70b27c36f420112152b26bd0ebe8fe/src/wp-includes/class-wp-scripts.php#L925C19-L980) method which is useful to refer to. 

Note this logic:

```php
// Handles with inline scripts attached in the 'after' position cannot be delayed.
if ( $this->has_inline_script( $handle, 'after' ) ) {
	return array();
}
```

So an enqueued script which is `async` falls back to blocking if it has an inline `after` script. If it has an inline `before` script, it can still be async.

Partytown scripts should essentially behave as if they are `async` scripts. In fact, it would be good to set the `async` strategy early on in addition to any Partytown-specific data flag on all Partytown scripts. If this `async` strategy "survives" the checks done by `filter_eligible_strategies` then the `script_loader_tag` filter can go ahead and replace the `async` attribute with `type="text/partytown"`. In this way, the `$before_script` and `$after_script` logic here can be eliminated.

Reply:
Thank you for the detailed explanation. It makes everything clear.

If a script is designed to run in a worker thread but is forced to remain asynchronous, should we consider implementing something to provide feedback about its behavior? Else it can confuse the user about the implementation.

Reply:
How about something like this:

```php

// Make sure that all partytown scripts are treated as if they are async.
add_filter(
	'print_scripts_array',
	static function ( array $script_handles ) {
		foreach ( array_intersect( $script_handles, wwo_get_partytown_handles() ) as $handle ) {
			wp_script_add_data( $handle, 'strategy', 'async' );
		}
	}
);

function wwo_update_script_type( string $tag, string $handle, string $src ): string {
	$partytown_handles = wwo_get_partytown_handles();

	// TODO: Make sure that previously the
	if ( in_array( $handle, $partytown_handles, true ) ) {
		$p = new WP_HTML_Tag_Processor( $tag );
		if ( $p->next_tag( array( 'tag_name' => 'SCRIPT' ) ) ) {
			if ( ! $p->get_attribute( 'async' ) ) {
				_doing_it_wrong(
					'wp_add_inline_script',
					esc_html(
						sprintf(
							/* translators: %s: script handle */
							__( 'Unable to offload "%s" script to a worker. Script will continue to load in the main thread.', 'web-worker-offloading' )
						)
					)
				);
			} else {
				$p->set_attribute( 'type', 'text/partytown' );
				$p->remove_attribute( 'async' );
				$tag = $p->get_updated_html();
			}
		}
	}

	return $tag;
}
```